File: homework0/rodrigo_canto/q1_UniqueSum.cpp
//Time Complexity: O(n) average-case
//Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

long long solve(const vector<int> &array){

    //Use an unordered set to keep track of distinct elements.
    unordered_set<long long> distinct_elements;

    for(long long value : array){
        distinct_elements.insert(value);
    }

    long long answer = 0;
    
    //Iterate through the set and add the value of the current element to the answer.
    for(auto element: distinct_elements){
        answer = answer + element;
    }

    return answer;
}

int main() {

    vector<vector<int>> cases;

    vector<int> case1 = {1, 10, 8, 3, 2, 5, 7, 2, -2, -1};
    vector<int> case2 = {4, 3, 3, 5, 7, 0, 2, 3, 8, 6};

    cases.emplace_back(case1);
    cases.emplace_back(case2);

    for(auto &test_case: cases){
        cout << solve(test_case) << "\n";
    }

    return 0;
}

//Time Spent: 8 minutes

File: homework0/bryan_zhao/q1_UniqueSum.py
#Time Complexity: O(n)
#Space Complexity: O(n)

def unique_sum(arr) -> int:
    total = 0
    unique_vals = set(arr)

    for val in unique_vals:
        total += val
    
    return total


test_case1 = [1, 10, 8, 3, 2, 5, 7, 2, -2, -1]
#Output: 33

test_case2 = [4, 3, 3, 5, 7, 0, 2, 3, 8, 6]
#Output: 35

print(unique_sum(test_case1))
print(unique_sum(test_case2))

#Time spent on unique_sum: 5 minutes